url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/CPU.java

```diff
@@ -0,0 +1,491 @@
+      } else if (charROM) {
+        return memory[rindex = adr | 0x10000];
+      } else {
+        return memory[rindex = adr];
+      }
+    } else {
+      return memory[rindex = adr];
+    }
+  }
+
+  // A byte is written directly to memory or to ioChips
+  protected final void writeByte(int adr, int data) {
+    cycles++;
+
+    schedule(cycles);
+    // Locking only on fetch byte...
+    // System.out.println("Writing byte at: " + Integer.toString(adr, 16)
+    // + " = " + data);
+    if (adr <= 1) {
+      memory[adr] = data;
+      int p = (memory[0] ^ 0xff) | memory[1];
+
+      kernalROM = ((p & 2) == 2); // Kernal on
+      basicROM = ((p & 3) == 3); // Basic on
+
+      charROM = ((p & 3) != 0) && ((p & 4) == 0);
+      // ioON is probably not correct!!! Check against table...
+      ioON = ((p & 3) != 0) && ((p & 4) != 0);
+
+      if (basicROM)
+        romFlag = 0xa000;
+      else if (kernalROM)
+        romFlag = 0xe000;
+      else
+        romFlag = 0x10000; // No Rom at all (Basic / Kernal)
+    }
+
+    adr &= 0xffff;
+    if (ioON && ((adr & 0xf000) == 0xd000)) {
+      // System.out.println("IO Write at: " + Integer.toString(adr, 16));
+      chips.performWrite(adr, data, cycles);
+    } else {
+      memory[windex = adr] = data;
+    }
+  }
+
+
+  private void fixRindex(int adr) {
+    // ROM/RAM address fix
+    if ((basicROM && ((adr & 0xe000) == 0xa000))
+        || (kernalROM && ((adr & 0xe000) == 0xe000))
+        || (charROM && ((adr & 0xf000) == 0xd000))) {
+      // Add ROM address for the read!
+      adr |= 0x10000;
+    }
+    rindex = adr;
+  }
+
+  public void poke(int address, int data) {
+    writeByte(address & 0xffff, data & 0xff);
+  }
+
+  public void patchROM(PatchListener list) {
+    this.list = list;
+
+    int pos = 0xf49e | 0x10000;
+    memory[pos++] = M6510Ops.JSR;
+    memory[pos++] = 0xd2;
+    memory[pos++] = 0xf5;
+
+    System.out.println("Patched LOAD at: " + Hex.hex2(pos));
+    memory[pos++] = LOAD_FILE;
+    memory[pos++] = M6510Ops.RTS;
+  }
+
+  public void runBasic() {
+    memory[631] = (int) 'R';
+    memory[632] = (int) 'U';
+    memory[633] = (int) 'N';
+    memory[634] = 13;// enter
+    memory[198] = 4; // length
+  }
+
+  public void enterText(String txt) {
+    System.out.println("Entering text into textbuffer: " + txt);
+    txt = txt.toUpperCase();
+    int len = txt.length();
+    int pos = 0;
+    for (int i = 0, n = len; i < n; i++) {
+      char c = txt.charAt(i);
+      if (c == '~')
+        c = 13;
+      memory[631 + pos] = c;
+      pos++;
+      if (pos == 5) {
+        memory[198] = pos;
+        pos = 0;
+        int tries = 5;
+        while (tries > 0 && memory[198] > 0) {
+          try {
+            Thread.sleep(50);
+          } catch (Exception e) {
+            e.printStackTrace();
+          }
+          tries--;
+          if (tries == 0) {
+            System.out.println("Buffer still full: " + memory[198]);
+          }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 208:218, 223:233.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.